Adjust add-on input validation

AhdraMeraliQB opened the pull request:
> **NOTE:** Kedro datasets are moving from `kedro.extras.datasets` to a separate `kedro-datasets` package in
> [`kedro-plugins` repository](https://github.com/kedro-org/kedro-plugins). Any changes to the dataset implementations
> should be done by opening a pull request in that repository.
## Description
<!-- Why was this PR created? -->

Following #3185, this PR ensures that invalid inputs to the add-on prompt fail immediately. Additionally, case sensitivity has been removed, whitespaces are now ignored, and selections larger than 9 are processed as invalid selections instead of improper formatting.

The case sensitivity changes have also been made to the add-on flag inputs (they were already whitespace robust prior to this PR).

## Development notes
<!-- What have you changed, and how has this been tested? -->


## Developer Certificate of Origin
We need all contributions to comply with the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). All commits must be signed off by including a `Signed-off-by` line in the commit message. [See our wiki for guidance](https://github.com/kedro-org/kedro/wiki/Guidelines-for-contributing-developers/).

If your PR is blocked due to unsigned commits, then you must follow the instructions under "Rebase the branch" on the GitHub Checks page for your PR. This will retroactively add the sign-off to all unsigned commits and allow the DCO check to pass.

## Checklist

- [x] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [x] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [x] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [x] Added tests to cover my changes
- [x] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

merelcht commented:
> Note: Values provided to the add-on flag remain case sensitive after these changes.

I would expect this to also have some processing to handle cases, the flows aren't entirely consistent, but I think this behaviour should be the same.

merelcht commented:
I also noticed that if I put in a negative number I still get the unhelpful message: `ValueError: invalid literal for int() with base 10: ''`

AhdraMeraliQB commented:
@merelcht 
> I would expect this to also have some processing to handle cases, the flows aren't entirely consistent, but I think this behaviour should be the same.

Done! Addressed in [a1fb0c6](https://github.com/kedro-org/kedro/pull/3263/commits/a1fb0c60813de674c0f1ab2994c4338abdf512be)


> I also noticed that if I put in a negative number I still get the unhelpful message: `ValueError: invalid literal for int() with base 10: ''`

Fixed in [906f63a](https://github.com/kedro-org/kedro/pull/3263/commits/906f63ae05a66aab59971d94468a54ee72b0bb48)

AhdraMeraliQB commented on tests/framework/cli/test_starters.py:
```diff
@@ -314,17 +314,6 @@ def test_parse_add_ons_invalid_range(input, capsys):
-    "input,first_invalid",
-    [("0,3,5", "0"), ("1,3,7", "7"), ("0-4", "0"), ("3-7", "7")],
-)
-def test_parse_add_ons_invalid_selection(input, first_invalid, capsys):
```
This test was removed as this validation step is no longer performed by `_parse_add_ons_input`

noklam commented on tests/framework/cli/test_starters.py:
```diff
@@ -906,7 +895,29 @@ def test_directory_flag_with_starter_alias(self, fake_kedro_cli):
     @pytest.mark.parametrize(
         "add_ons",
-        ["1", "2", "3", "4", "5", "6", "none", "2,3,4", "3-5", "all"],
+        [
+            "1",
+            "2",
+            "3",
+            "4",
+            "5",
+            "6",
+            "none",
+            "2,3,4",
+            "3-5",
+            "all",
+            "1, 2, 3",
+            "1,    2,   3",
+            "  1,  2, 3  ",
+            " 1-3 ",
+            "  1 - 3  ",
+            "ALL",
+            "All",
+            "AlL",
+            "NONE",
+            "None",
+            "nonE",
+        ],
```
Some of these tests are repetitive, can it be reduce some of these to one example. For example
```
            "ALL",
            "All",
            "AlL" 
 ```
can be reduced to one since `AlL` is probably the most weird case?

AhdraMeraliQB replied:
I do agree `AlL` is probably the strangest case, though I'd avoid simplifying to just either `ALL` or `All` as I'd want to ensure all small letters, all capitals, and a mix of the two are all processed correctly. `AlL` and `nonE` have both been dropped in [5ad8bab](https://github.com/kedro-org/kedro/pull/3263/commits/5ad8babe5a2f23eba1fe9ef30866f9cd6af254ff)

noklam reviewed (commented):
Looks good in general, leave a minor comment to reduce the test cases.

merelcht commented:
There's still a difference in user experience when I enter a negative number vs a positive invalid number:

![Screenshot 2023-11-03 at 13 50 04](https://github.com/kedro-org/kedro/assets/49397448/184fa2b9-0c36-4c00-8cfe-c40c38f5973f)

This might be fine, because `-` is also used for range here. @amandakys do you have any thoughts on this?

amandakys commented:
@merelcht Do you mean to say that a negative value should be accepted and auto corrected to be a positive value? Only when it falls within correct range, i.e we should validate the absolute value?

noklam commented:
I suggest test it after solving the conflicts. I suspect this maybe solved already as a side-effect of https://github.com/kedro-org/kedro/pull/3265. I introduced a new dictionary and use it everywhere, this avoids an addition rules because the condition it checking if value `in` the dictionary thus handle edge cases like negative value.

merelcht commented:
I found one more interesting edge case, what if you put a range that's way beyond the acceptable number, e.g. "5-12", now you'll get:

![Screenshot 2023-11-07 at 17 32 15](https://github.com/kedro-org/kedro/assets/49397448/85790447-ad23-4501-97d0-5442f635cf88)

Maybe I'm going too far into what users could reasonably enter, but I don't see why this shouldn't just throw: "'12' is not a valid selection.
Please select from the available add-ons: 1, 2, 3, 4, 5, 6, 7."

noklam commented:
@merelcht This is a reasonable expectation, I have made the changes and updated the tests.

SajidAlamQB commented:
Should we ignore white spaces? :

![image](https://github.com/kedro-org/kedro/assets/90610031/37397e45-2f87-445f-8524-4596176ff8a0)

![image](https://github.com/kedro-org/kedro/assets/90610031/4e73050f-6f2e-464f-8a6a-92082ef6fec3)

noklam commented:
@SajidAlamQB Did you test with latest branch? There are test cases covering that already and it seems working fine on my end.

![image](https://github.com/kedro-org/kedro/assets/18221871/db057bc3-eefc-4d30-883c-d60211f2cc9b)

SajidAlamQB commented:
> @SajidAlamQB Did you test with latest branch? There are test cases covering that already and it seems working fine on my end.

I created a new environment and it seemed to work there. Perhaps my old environment had some issues... 😅

merelcht reviewed (approved):
LGTM! 👍

SajidAlamQB reviewed (approved):
Thank you, @noklam!

noklam commented:
Thank you @AhdraMeraliQB for implementing the solution 😸 I just help to fix a test.
